packages/client/src/runtime/core/types/Utils.ts

```diff
@@ -94,3 +95,18 @@ export type Optional<O, K extends keyof any = keyof O> = {
 }
 
 export type Return<T> = T extends (...args: any[]) => infer R ? R : T
+
+// prettier-ignore
+export type RenameAndNestPayloadKeys<P> = {
+  [K in keyof P as K extends 'scalars' | 'objects' | 'composites' ? keyof P[K] : never]:
+    P[K] // we lift the value up with the same key name so we can flatten it later
+}
+
+// prettier-ignore
+export type PayloadToResult<P, O extends Record<any, any> = RenameAndNestPayloadKeys<P>> = {
+  [K in keyof O]?: O[K][K] extends any[]
+                   ? PayloadToResult<O[K][K][number]>[]
+                   : O[K][K] extends object
+                     ? PayloadToResult<O[K][K]>
+                    : O[K][K]
+}
```

Comment:
I'm unsure why `O[K][K]` is used here. Why the double-indexing?

Reply:
It's because `RenameAndNestPayloadKeys<P>` uses an exotic way to erase `scalars` and and `objects` keys. To do so:
```ts
type test = RenameAndNestPayloadKeys<{ objects: { user: { scalars: { id: string } } }>
```
Then yields
```ts
{ user: { user: { scalars: { id: string } } }
```
Then we remove the unneeded nesting
```
{ user: { user: { scalars: { id: string } } }['user']
```
Then you have a clean
```
{ user: { scalars: { id: string } }
```
Then repeat deeply

The goal and weirdness of this is to be able to merge top-level scalars, objects, composites as a single object. The other goal is to be able to only use a mapped type. We really want to avoid doing any intersection of objects, although that would be much simpler, because that would send the compiler to evaluate the types deeply. These are potentially recursive type definitions (eg. user: accounts[], account: user), so we really want to prevent intersections as they are inefficient.